### DISCUSSION TITLE
Plugin development: player joined hook?

### DISCUSSION THREAD

**Discussion by mixels** [Pryaxis/TShock#2090] [Q&A] [answered]
Looking for a hook or event that can fire when a player joins and gives me the name or id of the player. Is this a thing?

Also looking for the same for a player leaves event.

**Comment by dredhorse** [comment, marked as answer]
For join take a look here: https://github.com/dredhorse/tShockAutoTeam/blob/master/tShockAutoTeam/tShockAutoTeam.cs

should be similar for leave.

**Comment by mixels** [comment]
@dredhorse Thanks! I figured it out using a similar approach. For anyone else seeing this, there is a `ServerApi.Hooks.OnJoin` event that can be registered which returns `JoinEventArgs`, an object with only one property (`int Who`). That `int Who` property is the index of the player in `TShock.Players`. To get the corresponding TSPlayer object, use `TShock.Players[e.Who]`. The link from @dredhorse above shows a similar but different hook, `ServerApi.Hooks.NetGreetPlayer`.

From what I can tell, the difference between these is when they fire. `OnJoin` seems to fire immediately when the player's character name is resolved by the server. This happens even before the player enters the game world, in the "Connecting" screen shown in the game. `NetGreetPlayer`, on the other hand, fires when the player has fully loaded into the world.

Cheers and thanks!

**Reply by dredhorse** [reply]
Thanks for that information, documentation of those functions looks very thin to me, makes learning / coding harder than necessary.
